### PR TITLE
ClusterInfoProvider API and implementation

### PR DESCRIPTION
**Pull request by davidhoa**
Add API to provide basic cluster information. Initially this will just give a way to find the number of instances in a D2 cluster, based on scheme (http/https) and partitionid.  This new API will be part of Facilities. I provided a default implementation in Facilities, so this change is backwards compatible. This feature will allow dark cluster dispatchers to shape traffic to a dark canary cluster.

I'm not able to consolidate the tests because several of those variables are modified in many of the tests in different ways.

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public interface ClusterInfoProvider
+{
+  String HTTPS_SCHEME = "https";
+  int DEFAULT_PARTITION = 0;
+
+  /**
+   * Obtain d2 cluster count
+   * @return int
+   */
+  int getClusterCount(String clusterName, String scheme, int partitionId) throws ServiceUnavailableException;
```
Comment (davidhoa) [inline d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java]: what do you think about providing a fully async method? One idea is that I could provide the async as a base, and the blocking/sync as a convenience.

File: d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancer.java [hunk 2]
```diff
@@ -866,6 +869,30 @@ private void die(Callback<?> callback, String serviceName, String message)
     callback.onError(new ServiceUnavailableException(serviceName, message));
   }
 
+  @Override
+  public int getClusterCount(String clusterName, String scheme, int partitionId) throws ServiceUnavailableException
+  {
+    FutureCallback<Integer> clusterCountFutureCallback = new FutureCallback<>();
+    Runnable callback = () ->
```
Comment (zhang-chris) [inline d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancer.java]: We can inline this.

File: d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * ClusterInfoProvider provides a mechanism to access detailed cluster information from the D2 infrastructure.
+ *
+ * @author David Hoa
+ * @version $Revision: $
+ */
+public interface ClusterInfoProvider
+{
+  String HTTPS_SCHEME = "https";
+  int DEFAULT_PARTITION = 0;
```
Comment (zhang-chris) [inline d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java]: Are there public static variables for these already defined somewhere that we can use?

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java [hunk 4]
```diff
@@ -101,20 +101,27 @@
...
+import javafx.concurrent.Service;
 import javax.annotation.Nonnull;
 import org.apache.commons.io.FileUtils;
 import org.testng.Assert;
 import org.testng.annotations.AfterSuite;
 import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
+import static com.linkedin.d2.balancer.util.partitions.DefaultPartitionAccessor.*;
```
Comment (zhang-chris) [inline d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: expand imports

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java [hunk 5]
```diff
@@ -155,6 +162,87 @@ public void doOneTimeTearDown() throws IOException
...
+        // alter the partitions to check
+        new Object[] {0, 3, 0, 0, 0, 1},
+        new Object[] {3, 0, 0, 0, 0, 1},
+        new Object[] {1, 1, 0, 0, 0, 2},
+        new Object[] {0, 0, 0, 0, 0, 1},
+        // alter the partitions to add and check to match
+        new Object[] {0, 3, 0, 3, 1, 1},
+        new Object[] {3, 0, 3, 0, 1, 1},
+        new Object[] {1, 1, 1, 1, 2, 2},
+        new Object[] {0, 0, 0, 0, 1, 1}
```
Comment (zhang-chris) [inline d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: new Object[] is unnecessary.

File: d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancer.java (hunk 2 above)
Comment (davidhoa) [reply d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancer.java]: done.

File: d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java (hunk 3 above)
Comment (davidhoa) [reply d2/src/main/java/com/linkedin/d2/balancer/util/ClusterInfoProvider.java]: I switched to using DefaultPartitionAccessor.DEFAULT_PARTITION_ID, thanks for the suggestion. However, I did not find one for HTTPS_SCHEME. I added it to PropertyKeys, didn't find a better place for it.

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java (hunk 4 above)
Comment (davidhoa) [reply d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: done.

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java (hunk 5 above)
Comment (davidhoa) [reply d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: done

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java [hunk 6]
```diff
@@ -101,20 +100,27 @@
...
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNotNull;
 import static org.testng.Assert.assertTrue;
 import static org.testng.Assert.fail;
 
 public class SimpleLoadBalancerTest
 {
+  private static final String CLUSTER1_NAME = "cluster-1";
+  private static final String HTTP_SCHEME = "http";
+  private static final String HTTPS_SCHEME = "https";
```
Comment (zhang-chris) [inline d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: reference static variables

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java [hunk 7]
```diff
@@ -155,6 +161,87 @@ public void doOneTimeTearDown() throws IOException
...
+  public void testClusterInfoProvider(int numHttp, int numHttps, int expectedNumHttp, int expectedNumHttps,
+      int partitionIdForAdd, int partitionIdForCheck)
+      throws InterruptedException, ExecutionException, ServiceUnavailableException
+  {
+    Map<String, LoadBalancerStrategyFactory<? extends LoadBalancerStrategy>> loadBalancerStrategyFactories =
+        new HashMap<String, LoadBalancerStrategyFactory<? extends LoadBalancerStrategy>>();
+    Map<String, TransportClientFactory> clientFactories =
+        new HashMap<String, TransportClientFactory>();
+
+    MockStore<ServiceProperties> serviceRegistry = new MockStore<ServiceProperties>();
```
Comment (zhang-chris) [inline d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: -> new MockStore<>()

Here and everywhere else

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java (hunk 6 above)
Comment (davidhoa) [reply d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: done.

File: d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java (hunk 7 above)
Comment (davidhoa) [reply d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java]: I'm only gonna modify this function. Doing it for all others is a cleanup task that shouldn't be done with this RB, it'd obfuscate what I really added here.

**Pull request merged**